suseconnect/migration.go

```diff
@@ -0,0 +1,628 @@
+		//               break
+		//             end
+		//           end
+		//         end
+		//       end
+		//     end
+
+		msg = "Adding service " + service.Name
+		VerboseOut.Println(msg)
+		connect.MigrationAddService(service.URL, service.Name)
+
+		// store the base product version
+		if p.IsBase {
+			baseProductVersion = p.Version
+		}
+		if interrupted {
+			return baseProductVersion, fmt.Errorf("%s: %v", msg, ErrInterrupted)
+		}
+	}
+	return baseProductVersion, nil
+}
+
+// returns fs_inconsistent flag
+func applyMigration(migration connect.MigrationPath, quiet, verbose, nonInteractive bool, dupArgs []string) (bool, error) {
+	fsInconsistent := false
+	// TODO
+	//   zypp_backup(options[:root] ? options[:root]: "/")
+
+	if interrupted {
+		return fsInconsistent, fmt.Errorf("Preparing migration: %v", ErrInterrupted)
+	}
+
+	//   if system_products.detect { |p| p.identifier == "Leap" } &&
+	//      migration.detect { |p| p.identifier == "SLES" }
+	//      # bsc#1184237
+	//      print "Migration from Leap to SLES - disabling old repositories\n" unless options[:quiet]
+	//      SUSE::Connect::Migration::repositories.each do |repo|
+	//          SUSE::Connect::Migration::disable_repository repo[:name] if repo[:enabled] != 0
+	//      end
+	//   end
+
+	baseProductVersion, err := migrateSystem(migration)
+	if err != nil {
+		return fsInconsistent, err
+	}
+
+	echo := connect.SetSystemEcho(true)
+	if err := connect.RefreshRepos(baseProductVersion, true, false, false, false); err != nil {
+		return fsInconsistent, fmt.Errorf("Refresh of repositories failed: %v", err)
+	}
+	if interrupted {
+		return fsInconsistent, ErrInterrupted
+	}
+
+	err = connect.DistUpgrade(baseProductVersion, quiet, verbose, nonInteractive, dupArgs)
+	connect.SetSystemEcho(echo)
+	// TODO: export connect.zypperErrCommit (8)?
+	if err != nil && err.(connect.ZypperError).ExitCode == 8 {
+		fsInconsistent = true
+	}
+	if interrupted {
+		return fsInconsistent, ErrInterrupted
+	}
+
+	return fsInconsistent, err
+}
+
+func zypperDupArgs() []string {
+	// NOTE: "r" is not listed here as it shares values list with "repo"
+	wanted := map[string]struct{}{"auto-agree-with-licenses": {}, "l": {},
+		"allow-vendor-change": {}, "no-allow-vendor-change": {},
+		"debug-solver": {}, "recommends": {}, "no-recommends": {},
+		"replacefiles:": {}, "details": {}, "download": {},
+		"download-only": {}, "from": {}, "repo": {}}
+
+	args := []string{}
+
+	// special case (from original)
+	// mutually exclusive: allow-vendor-change/no-allow-vendor-change
+	avc := false
+	flag.Visit(func(f *flag.Flag) {
+		// skip not wanted
+		if _, found := wanted[f.Name]; !found {
+			return
+		}
+		// special case (update var and skip flag)
+		if strings.Contains(f.Name, "allow-vendor-change") {
+			avc = !strings.Contains(f.Name, "no-")
+			return
+		}
+		// multiArg? loop over values
+		if val, ok := f.Value.(*multiArg); ok {
+			for _, v := range *val {
+				args = append(args, "--"+f.Name, v)
+			}
```

Comment:
The previous `return`s are explicit "skip this flag" cases. Here we don't know if some more processing will be added after this `if` or not. I'd rather keep it that way to avoid future mistakes.

Reply:
ok makes sense